[windowing/gbm] Select best CRTC+planes option

sigmaris opened the pull request:
## Description

On a platform with multiple CRTCs connected to the selected encoder, instead of using the default CRTC currently driving the encoder, get the availability of video and GUI planes on each CRTC, then select the best out of all found CRTC+planes options.
v2: smarter CRTC selection.
v3: iterate over CRTC options and use the best "scoring" one instead of sorting the list. Follow coding guidelines on alignment & spacing.
v4: base on https://github.com/lrusak/xbmc/commit/86000d34446ecc89f9e170cfcee8502f22ef352c and disable original CRTC in atomic modeset if choosing a different one.

## Motivation and Context

On an RK3399 SoC (Rockpro64 board) with the mainline kernel, there are two CRTCs that can be used to drive HDMI, the "little VOP" or vopl with only one primary plane and one cursor plane, and the "big VOP" or vopb which supports several planes. The mainline kernel configures HDMI to be driven by vopl by default.

Kodi doesn't use cursor planes, and the single primary plane supports NV12 format which is the only criteria used when searching for `KODI_VIDEO_PLANE`, so in `FindPlanes()` the primary plane is used for the `KODI_VIDEO_PLANE` and then no GUI plane can be found. Kodi segfaults later in `FindPlanes()` when it tries to get the properties of the null GUI plane. Kodi should be smarter in CRTC+planes selection and pick the more suitable CRTC.

## How Has This Been Tested?

Tested on a 5.2-rc1 kernel on the Rockpro64 board - with this patch kodi-gbm will start, scan the available CRTCs, switch HDMI to be driven by vopb and display the normal UI on an attached HDMI monitor. Previously it would exit with a segmentation fault.
Also tested on Raspberry Pi 2 using the vc4 DRM and Mesa GBM driver.

## Types of change
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

sigmaris commented:
[drm_info](https://github.com/ascent12/drm_info) output for /dev/dri/card0 on this board, if it's of interest: https://gist.github.com/sigmaris/54bc5702dc9219bce367838e7cfac562

lrusak commented:
@Kwiboo can you help @sigmaris with the rockchip stuff? I'm not sure how you are running the mainline kernel on your rockchip boards.

Kwiboo commented:
@sigmaris vopl is probably routed to your hdmi connector, on LibreELEC we disable vopl because it is limiting resolution and planes, try https://github.com/Kwiboo/linux-rockchip/commit/5ff51ab11cc0821aedf0ea62af934a5e1954c5d3 or similar and see kodi finds proper video and gui planes

sigmaris commented:
@Kwiboo Yes, I guess CRTC ID 34 is the vopl and CRTC ID 39 is vopb. I could use https://github.com/Kwiboo/linux-rockchip/commit/5ff51ab11cc0821aedf0ea62af934a5e1954c5d3 to disable vopl but as you say in the commit subject it's a bit of a hack, and I'd like to try and get Kodi running eventually on a standard-as-possible kernel configuration. What would be your opinion on a patch for Kodi that changed from selecting just the current active CRTC for the conector+encoder, to considering all available CRTCs and selecting the "best" one (where "best" is defined as having video and GUI planes available)?

Kwiboo commented:
I do agree Kodi should be smarter in crtc+plane selection process and should not crash if a video plane is found and no gui plane. Can you extend this PR with a smarter crtc selection? I this your definition of "best" crtc is valid, first crtc having both video and gui plane or fall back to first crtc having a valid gui plane.

Disabling vopl is an easy workaround but it also fixed an issue on vendor kernel where 4k resolutions got filtered out for vopb based on vopl size limits even when Kodi selected CRTC ID for vopb. Since 4k is not working on rk mainline I decided to keep using the vopl disable workaround for now.

Do not be scared by the `HACK` prefix, I mainly use it to sort out what patches won't make it upstream but that I plan on carrying in LibreELEC, see https://github.com/Kwiboo/linux-rockchip/compare/rockchip-5.1-work-in-progress-v5.1.5...rockchip-5.1-patches-for-libreelec-v5.1.5 for other `HACK` patches.

sigmaris commented:
I've updated this with an implementation that looks for all current or inactive CRTCs for the chosen encoder/connector, looks for the planes for each CRTC and tries to pick the best CRTC+planes option out of them all. With the patch, on the Rockpro64 Kodi now selects vopb, with video + gui planes, and switches the CRTC driving the HDMI connector from vopl to vopb.

One thing I haven't managed to test is to get Kodi to render to the video plane and then blend it with the GUI plane. I'm assuming that requires a hardware decoder with DRM PRIME support. Kodi doesn't want to use the Rockchip VPU on the Rockpro64 with the mainline kernel (I guess missing FFMpeg v4l2-request patches). I also tried with a Raspberry Pi 2 with the "bcm2835-codec" v4l2-m2m driver but couldn't get it to decode H.264 and display it direct-to-plane. So far everything I've tested uses software decoding and displays it on the GUI plane.

Kwiboo commented:
Thanks for working on this. You are correct in that you need rk vpu hw decoder (mpeg-2 decoder now merged in latest media tree and should land in linux 5.3, and [ffmpeg patches for adding support for v4l2 request api](https://github.com/Kwiboo/FFmpeg/commits/v4l2-request-hwaccel)) or a work-in-progress kodi patch that can render sw decoded video on a video plane, [early version here](https://github.com/Kwiboo/LibreELEC.tv/blob/rockchip-5.x/projects/Rockchip/patches/kodi/kodi-PR16102.patch).

I will test and review this PR while I finish up on the sw decode work, hopefully in a few days.

lrusak commented:
Thanks for a the work on this. I have to say though it seems like this is way more complex than it needs to be. Adding 4 extra methods just to select a CRTC seems overkill to me.

Rechi commented on xbmc/windowing/gbm/DRMUtils.cpp:
```diff
@@ -274,33 +275,158 @@ bool CDRMUtils::FindEncoder()
+  // sorts crtcs from least to most suitable
+  crtcs.sort(CDRMUtils::CompareCrtcOptions);
+
+  struct crtc_option &best_crtc = crtcs.back();
```
left align `&`

Rechi commented on xbmc/windowing/gbm/DRMUtils.cpp:
```diff
@@ -274,33 +275,158 @@ bool CDRMUtils::FindEncoder()
+  return true;
+}
+
+bool CDRMUtils::CompareCrtcOptions(const struct crtc_option &l, const struct crtc_option &r)
```
left align `&`

Rechi commented on xbmc/windowing/gbm/DRMUtils.cpp:
```diff
@@ -274,33 +275,158 @@ bool CDRMUtils::FindEncoder()
   {
-    m_crtc->crtc = drmModeGetCrtc(m_fd, m_drm_resources->crtcs[i]);
-    if(m_crtc->crtc->crtc_id == m_encoder->encoder->crtc_id)
+    if(m_encoder->encoder->possible_crtcs & (1 << i))
```
space missing between if and `(`

Rechi commented on xbmc/windowing/gbm/DRMUtils.cpp:
```diff
@@ -274,33 +275,158 @@ bool CDRMUtils::FindEncoder()
+      }
+
+      // If this CRTC is available and has suitable planes, add it to the list
+      if((crtc_opt.crtc_obj->crtc->crtc_id == m_encoder->encoder->crtc_id || !CrtcIsActive(crtc_opt.crtc_obj))
```
space missing between if and `(`

Rechi commented on xbmc/windowing/gbm/DRMUtils.cpp:
```diff
@@ -274,33 +275,158 @@ bool CDRMUtils::FindEncoder()
+  return crtcs;
+}
+
+bool CDRMUtils::FindCrtcOptionPlanes(drmModePlaneResPtr resources, struct crtc_option &crtc_opt)
```
left align `&`

Rechi commented on xbmc/windowing/gbm/DRMUtils.cpp:
```diff
@@ -274,33 +275,158 @@ bool CDRMUtils::FindEncoder()
+  return crtc_opt.gui_plane != nullptr;
+}
+
+bool CDRMUtils::CrtcIsActive(struct crtc *crtc)
```
left align `*`

Rechi commented on xbmc/windowing/gbm/DRMUtils.cpp:
```diff
@@ -348,13 +474,13 @@ bool CDRMUtils::SupportsFormat(drmModePlanePtr plane, uint32_t format)
 }
 
-drmModePlanePtr CDRMUtils::FindPlane(drmModePlaneResPtr resources, int crtc_index, int type)
+drmModePlanePtr CDRMUtils::FindPlane(drmModePlaneResPtr resources, const struct crtc_option &crtc_opt, int type)
```
left align `&`

Rechi commented on xbmc/windowing/gbm/DRMUtils.cpp:
```diff
@@ -635,7 +743,7 @@ bool CDRMUtils::InitDrm()
     }
 
-    if(!FindPlanes())
+    if(!GetPlanesProperties())
```
space between if and `(` is missing

Rechi commented on xbmc/windowing/gbm/DRMUtils.h:
```diff
@@ -14,6 +14,8 @@
 #include <map>
 #include <vector>
+#include <list>
+#include <tuple>
```
wrong order, list before map and tuple before vector

Rechi commented on xbmc/windowing/gbm/DRMUtils.h:
```diff
@@ -90,6 +92,28 @@ struct drm_fb
 
+struct crtc_option
+{
+  struct crtc *crtc_obj;
```
left align `*`

Rechi commented on xbmc/windowing/gbm/DRMUtils.h:
```diff
@@ -132,6 +156,7 @@ class CDRMUtils
   struct connector *m_connector = nullptr;
   struct encoder *m_encoder = nullptr;
   struct crtc *m_crtc = nullptr;
+  struct crtc *m_orig_crtc = nullptr;
```
left align `*`

Rechi commented on xbmc/windowing/gbm/DRMUtils.h:
```diff
@@ -141,13 +166,17 @@ class CDRMUtils
 private:
   static bool SupportsFormat(drmModePlanePtr plane, uint32_t format);
-  drmModePlanePtr FindPlane(drmModePlaneResPtr resources, int crtc_index, int type);
+  drmModePlanePtr FindPlane(drmModePlaneResPtr resources, const struct crtc_option &crtc_opt, int type);
```
left align `&`

Rechi commented on xbmc/windowing/gbm/DRMUtils.h:
```diff
@@ -141,13 +166,17 @@ class CDRMUtils
   bool FindEncoder();
   bool FindCrtc();
-  bool FindPlanes();
+  static bool CompareCrtcOptions(const struct crtc_option &l, const struct crtc_option &r);
```
left align `&`

Rechi commented on xbmc/windowing/gbm/DRMUtils.h:
```diff
@@ -141,13 +166,17 @@ class CDRMUtils
-  bool FindPlanes();
+  static bool CompareCrtcOptions(const struct crtc_option &l, const struct crtc_option &r);
+  std::list<struct crtc_option> FindCrtcOptions();
+  bool CrtcIsActive(struct crtc *crtc);
```
left align `*`

Rechi commented on xbmc/windowing/gbm/DRMUtils.h:
```diff
@@ -141,13 +166,17 @@ class CDRMUtils
+  std::list<struct crtc_option> FindCrtcOptions();
+  bool CrtcIsActive(struct crtc *crtc);
+  bool GetPlanesProperties();
+  bool FindCrtcOptionPlanes(drmModePlaneResPtr resources, struct crtc_option &crtc_opt);
```
left align `&`

sigmaris commented:
@lrusak I wasn't sure about checking if the CRTC was active before adding it to the list of options; in the case I'm testing with, the vopb CRTC is inactive, but in the case where the preferred CRTC is active and driving a different connector, I wasn't sure if Kodi should try and "steal" it. If it's OK to use other, active CRTCs, I could remove `CrtcIsActive()` and the call to it.

The selection of best CRTC could also be changed to get a score for each `crtc_option` and pick the top scoring one; that could avoid the `CompareCrtcOptions()` method - probably more LOC overall but fewer methods & less use of stdlib. Beyond that, `GetPlaneProperties()` could be moved into `FindCrtc()` but that's just shuffling code around. I don't think iterating all the possible CRTCs, getting their planes and then selecting the best one can really be avoided, so I don't think it can be cut down much more.

@Rechi thanks for the review, I'll update the alignment and spaces in the next version.

phunkyfish commented:
@sigmaris do you want to progress further?

@lrusak do you have any further review comments?

sigmaris commented:
@phunkyfish Yes, I think this is still relevant and I'd like to get some kind of fix for this issue merged if possible.

I tried compiling and running latest master on my RockPro64 yesterday, it still runs into the segmentation fault described above. I rebased this patch on master and applied it, and it still works to fix the issue.

Should I just push the rebased commit to this PR?

lrusak commented:
@sigmaris I appreciate the work on this and I agree it would be nice to make it work on the mainline kernel without any patches.

I plan to remove the 10bit plane support as it's really not needed, that should simplify the plane selection.

Then I'll try and review this to see if we can make it fit 😃

lrusak commented:
@sigmaris ok I pushed the 10bit plane removal so we can work on this.

One thing I would like to say right away though is that I don't like having a weighting system. In my opinion we either have something we need or we don't. We should look for exactly what we need and if we don't find that then we need a safe fallback option.

sigmaris commented:
@lrusak Great, I will revisit this in the next few days.

> One thing I would like to say right away though is that I don't like having a weighting system. In my opinion we either have something we need or we don't. We should look for exactly what we need and if we don't find that then we need a safe fallback option.

When I last looked at this, Kodi "would like" but didn't "need" a separate video plane, as a fallback choice it could work with a single GUI plane.

So taking the perspective of "look for exactly what we ~need~ would like", I'm thinking Kodi should look through all CRTC-options for a CRTC that supports separate video & GUI planes first, and then if that's not found, for "a safe fallback option" it should look through all CRTC-options for a CRTC that supports a GUI plane alone. (And if that's finally not found, it should fail overall, as before)

lrusak commented:
> When I last looked at this, Kodi "would like" but didn't "need" a separate video plane, as a fallback choice it could work with a single GUI plane.

Yes, the safe option is just a single plane for the gui that supports XRGB8888


> So taking the perspective of "look for exactly what we ~need~ would like", I'm thinking Kodi should look through all CRTC-options for a CRTC that supports separate video & GUI planes first, and then if that's not found, for "a safe fallback option" it should look through all CRTC-options for a CRTC that supports a GUI plane alone. (And if that's finally not found, it should fail overall, as before)

I think the simple option would be to enumerate all available crtc's for a connector, then when looking for planes just loop through them until we find both planes that both support the crtc. If we don't find both planes we will still likely have an available GUI plane when looking.

lrusak committed:
[gbm] rework plane selection based on crtcs

lrusak commented:
ok I've looked into this a little bit. What do you think about something like this? https://github.com/lrusak/xbmc/commit/86000d34446ecc89f9e170cfcee8502f22ef352c

I've tested on two systems, one that supports NV12 and one that doesn't

supports NV12
```
2020-04-14 14:57:49.349 T:720   DEBUG: CDRMUtils::FindConnector - found connector: 143
2020-04-14 14:57:49.349 T:720   DEBUG: CDRMUtils::FindEncoder - found encoder: 142
2020-04-14 14:57:49.350 T:720   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 67
2020-04-14 14:57:49.350 T:720   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 104
2020-04-14 14:57:49.350 T:720   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 141
2020-04-14 14:57:49.350 T:720   DEBUG: CDRMUtils::FindPlane - found video plane 31
2020-04-14 14:57:49.351 T:720   DEBUG: CDRMUtils::FindPlane - found gui plane 39
2020-04-14 14:57:49.351 T:720   DEBUG: CDRMUtils::FindPlanes - using video plane 31
2020-04-14 14:57:49.351 T:720   DEBUG: CDRMUtils::FindPlanes - using gui plane 39
```

doesn't support NV12
```
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindConnector - found connector: 79
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindEncoder - found encoder: 78
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 47
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 49
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 51
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 53
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 55
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 57
2020-04-14 14:43:59.161 T:1292237 WARNING: CDRMUtils::FindPlane - could not find video plane for crtc index 0
2020-04-14 14:43:59.161 T:1292237   DEBUG: CDRMUtils::FindPlane - found gui plane 45
2020-04-14 14:43:59.162 T:1292237 WARNING: CDRMUtils::FindPlane - could not find video plane for crtc index 1
2020-04-14 14:43:59.162 T:1292237   DEBUG: CDRMUtils::FindPlane - found gui plane 44
2020-04-14 14:43:59.162 T:1292237 WARNING: CDRMUtils::FindPlane - could not find video plane for crtc index 2
2020-04-14 14:43:59.162 T:1292237   DEBUG: CDRMUtils::FindPlane - found gui plane 43
2020-04-14 14:43:59.162 T:1292237 WARNING: CDRMUtils::FindPlane - could not find video plane for crtc index 3
2020-04-14 14:43:59.162 T:1292237   DEBUG: CDRMUtils::FindPlane - found gui plane 42
2020-04-14 14:43:59.163 T:1292237 WARNING: CDRMUtils::FindPlane - could not find video plane for crtc index 4
2020-04-14 14:43:59.163 T:1292237   DEBUG: CDRMUtils::FindPlane - found gui plane 41
2020-04-14 14:43:59.163 T:1292237 WARNING: CDRMUtils::FindPlane - could not find video plane for crtc index 5
2020-04-14 14:43:59.163 T:1292237   DEBUG: CDRMUtils::FindPlane - found gui plane 40
2020-04-14 14:43:59.163 T:1292237   DEBUG: CDRMUtils::FindPlanes - using gui plane 45
```

Maybe you can give this a test and see if it works on your system?

sigmaris commented:
> Maybe you can give this a test and see if it works on your system?

Thanks! - The finding CRTC and planes seems to work well:
```
2020-04-15 21:38:42.947 T:20961   DEBUG: CDRMUtils::FindConnector - found connector: 46
2020-04-15 21:38:42.953 T:20961   DEBUG: CDRMUtils::OpenDrm - opened device: /dev/dri/card0 using module: rockchip
2020-04-15 21:38:43.058 T:20961   DEBUG: CDRMUtils::FindConnector - found connector: 46
2020-04-15 21:38:43.058 T:20961   DEBUG: CDRMUtils::FindEncoder - found encoder: 45
2020-04-15 21:38:43.058 T:20961   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 35
2020-04-15 21:38:43.059 T:20961   DEBUG: CDRMUtils::FindCrtcs - found possible crtc: 40
2020-04-15 21:38:43.059 T:20961   DEBUG: CDRMUtils::FindPlane - found video plane 31
2020-04-15 21:38:43.060 T:20961 WARNING: CDRMUtils::FindPlane - could not find gui plane for crtc index 0
2020-04-15 21:38:43.060 T:20961   DEBUG: CDRMUtils::FindPlane - found video plane 36
2020-04-15 21:38:43.061 T:20961   DEBUG: CDRMUtils::FindPlane - found gui plane 41
2020-04-15 21:38:43.061 T:20961   DEBUG: CDRMUtils::FindPlanes - using video plane 36
2020-04-15 21:38:43.061 T:20961   DEBUG: CDRMUtils::FindPlanes - no drm modifiers present for the video plane
2020-04-15 21:38:43.061 T:20961   DEBUG: CDRMUtils::FindPlanes - using gui plane 41
2020-04-15 21:38:43.061 T:20961   DEBUG: CDRMUtils::FindPlanes - no drm modifiers present for the gui plane
2020-04-15 21:38:43.061 T:20961 WARNING: CDRMUtils::InitDrm - failed to set drm master, will try to authorize instead: Permission denied
2020-04-15 21:38:43.062 T:20961  NOTICE: CDRMUtils::InitDrm - successfully authorized drm magic
2020-04-15 21:38:43.063 T:20961   DEBUG: CDRMAtomic::InitDrm - initialized atomic DRM
2020-04-15 21:38:43.715 T:20961   DEBUG: CWinSystemGbm::InitWindowSystem - initialized DRM
```

The only hiccup is later, the first DRM atomic commit to set the video mode fails:

```
2020-04-15 21:38:44.027 T:20961   DEBUG: CDRMAtomic::FlipPage - Execute modeset at next commit
2020-04-15 21:38:44.028 T:20961   ERROR: CDRMAtomic::DrmAtomicCommit - test commit failed: Invalid argument
```

As far as I remember, this is probably because the original CRTC outputting to that connector has to be disabled when switching to use a different CRTC, as I had to do in this commit: https://github.com/sigmaris/xbmc/commit/0962551d5a3a4a2d4bdf53836ce31543a5e5ea75#diff-b75bd81bb7915d3cac61a75cf7949b3fR42-R53

lrusak commented:
Interesting! would you mind trying it with that and seeing if you can get it to work?

sigmaris commented:
@lrusak Adding in https://github.com/sigmaris/xbmc/commit/9f4ea36279834aceaddffc2a88f509effb32322f gets things working for me (Kodi uses CRTC 40 with separate video & GUI planes, setting different video modes works)

lrusak commented:
@sigmaris awesome! Do you want to pick my patch and add your other commit here? They can get squashed together under your name :)

Also make sure to check the formatting with clang-format please! :wink:

sigmaris committed:
[windowing/gbm] Set original CRTC as inactive if not used

In the case of choosing a non-default CRTC, we need to explicitly disable the
formerly used CRTC for the connector when using the DRM Atomic API to modeset
using the new CRTC, otherwise the test commit will fail with EINVAL. This is not
necessary for the DRM Legacy API as it will disable the former CRTC implicitly.

lrusak commented:
@sigmaris would you mind applying this diff to your commit to comply with our formatting guidlines? Thanks!

https://jenkins.kodi.tv/job/BuildMulti-PR/15618/artifact/PR16237-0002-d2d7a1e070-windowing-gbm-Set-original-CRTC-as.diff

lrusak reviewed (approved):
Thanks for all your hard work on this!

sigmaris commented:
Thanks @lrusak !
